### PR TITLE
JDK-8312617: SIGSEGV in ConnectionGraph::verify_ram_nodes

### PR DESCRIPTION
**Pull request by JohnTortugo**
- Return early from `verify_ram_nodes` if compilation is already failing.
- Add back check for `failing()` after `eliminate_macro_nodes()`.
- Print additional diagnostic information when an unexpected user of RAM is encountered.

Tested with tier1-3 on Linux x64.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312617](https://bugs.openjdk.org/browse/JDK-8312617): SIGSEGV in ConnectionGraph::verify_ram_nodes (**Bug** - P2)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [b55cba76](https://git.openjdk.org/jdk/pull/15048/files/b55cba76c4348c9becd66d5dec1ea1e404bf3f5a)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15048/head:pull/15048` \
`$ git checkout pull/15048`

Update a local copy of the PR: \
`$ git checkout pull/15048` \
`$ git pull https://git.openjdk.org/jdk.git pull/15048/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15048`

View PR using the GUI difftool: \
`$ git pr show -t 15048`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15048.diff">https://git.openjdk.org/jdk/pull/15048.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15048#issuecomment-1652626965)

### CODE REVIEW

File: src/hotspot/share/opto/escape.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/escape.cpp:508]: 1. You don't need to check `ReduceAllocationMerges` because `do_reduce_allocation_merges` is set by its value.
2. Setting `do_reduce_allocation_merges` in `C2Compiler::compile_method()` should also check `EliminateAllocations` flag. Then you don't need to check this flag here.

File: src/hotspot/share/opto/escape.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/escape.cpp:742]: 1. move call to this `verify_ram_nodes()` method after `failed()` check in `compile.cpp` so you don need to check it here.
2. Check `_compile->do_reduce_allocation_merges()` instead of flag and move the check at the first line of this method.

**Review by vnkozlov — COMMENTED** [review]
Use `_compile->do_reduce_allocation_merges()` instead of `ReduceAllocationMerges` in all places in `escape.cpp`.   You removed one in `adjust_scalar_replaceable_state()`. I see an other at line 51 (`_nodes` field allocation) and at line 401 (where you don't need it - `reducible_merges.size() > 0` check is enough)

**Comment by JohnTortugo** [comment]
Thanks for reviewing @vnkozlov

**Review by vnkozlov — APPROVED** [review]
Looks good.
You need second review.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me otherwise. I'll run some testing and report back once it passed.

**Review by TobiHartmann — APPROVED** [review]
All tests passed.

**Comment by JohnTortugo** [comment]
Thank you for testing/reviewing @TobiHartmann / @vnkozlov

**Pull request closed without merging**
